From 1e280e8b0ca43af6612f3e7f79bc3e419e4c5cca Mon Sep 17 00:00:00 2001
From: liyong4x <yongx.a.li@intel.com>
Date: Tue, 12 Jun 2018 04:09:43 -0400
Subject: [PATCH 088/129] media: intel-ipu4: pdata: add ar0231at configurations

Message for Open Source:
Ar0231at need to be supported as a subdevice of max9286.

Message for Internal:
Ar0231at need to be supported as a subdevice of max9286.

[Issue/Feature] Add ar0231at configurations.

[Root Cause/Changes]
1. Configure the I2C clock for max9286 and ar0231at;
2. Register at0231at as a subdevice of max9286.

Change-Id: If1c20cbf2d2b37c35ad735c3674ec96cfad26b1a
Tracked-On: #JIIAP-475
Signed-off-by: liyong4x <yongx.a.li@intel.com>
Signed-off-by: Meng Wei <wei.meng@intel.com>
---
 drivers/media/platform/intel/ipu4-bxt-p-pdata.c | 87 +++++++++++++++++++++++++
 1 file changed, 87 insertions(+)

diff --git a/drivers/media/platform/intel/ipu4-bxt-p-pdata.c b/drivers/media/platform/intel/ipu4-bxt-p-pdata.c
index 52cab10..334b105 100755
--- a/drivers/media/platform/intel/ipu4-bxt-p-pdata.c
+++ b/drivers/media/platform/intel/ipu4-bxt-p-pdata.c
@@ -10,6 +10,7 @@
 #include <media/ipu-isys.h>
 #include <media/crlmodule.h>
 #include <media/ti964.h>
+#include <media/max9286.h>
 #include "ipu.h"
 #include <media/dw9714.h>
 #define GPIO_BASE		422
@@ -952,6 +953,87 @@
 };
 #endif
 
+#ifdef CONFIG_INTEL_IPU4_AR0231AT
+#define AR0231AT_LANES            4
+#define AR0231ATA_I2C_ADDRESS      0x11
+#define AR0231ATB_I2C_ADDRESS      0x12
+#define AR0231ATC_I2C_ADDRESS      0x13
+#define AR0231ATD_I2C_ADDRESS      0x14
+
+static struct crlmodule_platform_data ar0231at_pdata = {
+	.lanes = AR0231AT_LANES,
+	.ext_clk = 24000000,
+	.op_sys_clock = (uint64_t[]){ 264000000 },
+	.module_name = "AR0231AT",
+};
+#endif
+
+#if IS_ENABLED(CONFIG_VIDEO_MAX9286)
+#define DS_MAX9286_LANES                4
+#define DS_MAX9286_I2C_ADAPTER          4
+#define DS_MAX9286_I2C_ADDRESS          0x48
+
+static struct ipu_isys_csi2_config max9286_csi2_cfg = {
+	.nlanes = DS_MAX9286_LANES,
+	.port = 4,
+};
+
+static struct max9286_subdev_i2c_info max9286_subdevs[] = {
+#ifdef CONFIG_INTEL_IPU4_AR0231AT
+		{
+			.board_info = {
+				.type = CRLMODULE_NAME,
+				.addr = AR0231ATA_I2C_ADDRESS,
+				.platform_data = &ar0231at_pdata,
+			},
+			.i2c_adapter_id = DS_MAX9286_I2C_ADAPTER,
+		},
+		{
+			.board_info = {
+				.type = CRLMODULE_NAME,
+				.addr = AR0231ATB_I2C_ADDRESS,
+				.platform_data = &ar0231at_pdata,
+			},
+			.i2c_adapter_id = DS_MAX9286_I2C_ADAPTER,
+		},
+		{
+			.board_info = {
+				.type = CRLMODULE_NAME,
+				.addr = AR0231ATC_I2C_ADDRESS,
+				.platform_data = &ar0231at_pdata,
+			},
+			.i2c_adapter_id = DS_MAX9286_I2C_ADAPTER,
+		},
+		{
+			.board_info = {
+				.type = CRLMODULE_NAME,
+				.addr = AR0231ATD_I2C_ADDRESS,
+				.platform_data = &ar0231at_pdata,
+			},
+			.i2c_adapter_id = DS_MAX9286_I2C_ADAPTER,
+		},
+#endif
+};
+
+static struct max9286_pdata max9286_pdata = {
+	.subdev_info = max9286_subdevs,
+	.subdev_num = ARRAY_SIZE(max9286_subdevs),
+	.reset_gpio = GPIO_BASE + 63,
+};
+
+static struct ipu_isys_subdev_info max9286_sd = {
+	.csi2 = &max9286_csi2_cfg,
+	.i2c = {
+		.board_info = {
+			.type = "max9286",
+			.addr = DS_MAX9286_I2C_ADDRESS,
+			.platform_data = &max9286_pdata,
+		},
+		.i2c_adapter_id = DS_MAX9286_I2C_ADAPTER,
+	}
+};
+#endif
+
 /*
  * Map buttress output sensor clocks to sensors -
  * this should be coming from ACPI
@@ -967,6 +1049,8 @@
 	{ CLKDEV_INIT("0-0010", NULL, NULL), "OSC_CLK_OUT0" },
 	{ CLKDEV_INIT("2-000e", NULL, NULL), "OSC_CLK_OUT0" },
 	{ CLKDEV_INIT("4-000e", NULL, NULL), "OSC_CLK_OUT1" },
+	{ CLKDEV_INIT("0-0048", NULL, NULL), "OSC_CLK_OUT0" },
+	{ CLKDEV_INIT("4-0048", NULL, NULL), "OSC_CLK_OUT1" },
 	{ CLKDEV_INIT(NULL, NULL, NULL), NULL }
 };
 
@@ -1029,6 +1113,9 @@
 #ifdef CONFIG_INTEL_IPU4_OV2775
 		&ov2775_crl_sd,
 #endif
+#if IS_ENABLED(CONFIG_VIDEO_MAX9286)
+		&max9286_sd,
+#endif
 		NULL,
 	},
 	.clk_map = clk_mapping,
-- 
1.9.1

